File: app/src/main/java/us/shandian/giga/io/FileStream.java
package us.shandian.giga.io;

import androidx.annotation.NonNull;

import org.schabi.newpipe.streams.io.SharpStream;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;

/**
 * @author kapodamy
 */
public class FileStream extends SharpStream {

    public RandomAccessFile source;

    public FileStream(@NonNull File target) throws FileNotFoundException {
        this.source = new RandomAccessFile(target, "rw");
    }

    public FileStream(@NonNull String path) throws FileNotFoundException {
        this.source = new RandomAccessFile(path, "rw");
    }

    @Override
    public int read() throws IOException {
        return source.read();
    }

    @Override
    public int read(byte b[]) throws IOException {
        return source.read(b);
    }

    @Override
    public int read(byte b[], int off, int len) throws IOException {
        return source.read(b, off, len);
    }

    @Override
    public long skip(long pos) throws IOException {
        return source.skipBytes((int) pos);
    }

    @Override
    public long available() {
        try {
            return source.length() - source.getFilePointer();
        } catch (IOException e) {
            return 0;
        }
    }

    @Override
    public void close() {
        if (source == null) return;
        try {
            source.close();
        } catch (IOException err) {
            // nothing to do
        }
        source = null;
    }

    @Override
    public boolean isClosed() {
        return source == null;
    }

    @Override
    public void rewind() throws IOException {
        source.seek(0);
    }

    @Override
    public boolean canRewind() {
        return true;
    }

    @Override
    public boolean canRead() {
        return true;
    }

    @Override
    public boolean canWrite() {
        return true;
    }

    @Override
    public boolean canSeek() {
        return true;
    }

    @Override
    public boolean canSetLength() {
        return true;
    }

    @Override
    public void write(byte value) throws IOException {
        source.write(value);
    }

    @Override
    public void write(byte[] buffer) throws IOException {
        source.write(buffer);
    }

    @Override
    public void write(byte[] buffer, int offset, int count) throws IOException {
        source.write(buffer, offset, count);
    }

    @Override
    public void setLength(long length) throws IOException {
        source.setLength(length);
    }

    @Override
    public void seek(long offset) throws IOException {
        source.seek(offset);
    }

    @Override
    public long length() throws IOException {
        return source.length();
    }
}

File: app/src/main/java/us/shandian/giga/io/FileStream.kt
package us.shandian.giga.io

import java.io.File
import java.io.FileNotFoundException
import okio.IOException
import java.io.RandomAccessFile
import org.schabi.newpipe.streams.io.SharpStream

/**
 * @author kapodamy
 */
class FileStream : SharpStream {

    var source: RandomAccessFile? = null

    @Throws(FileNotFoundException::class)
    constructor(target: File) {
        this.source = RandomAccessFile(target, "rw")
    }

    @Throws(FileNotFoundException::class)
    constructor(path: String) {
        this.source = RandomAccessFile(path, "rw")
    }

    @Throws(IOException::class)
    override fun read(): Int {
        return source!!.read()
    }

    @Throws(IOException::class)
    override fun read(buffer: ByteArray): Int {
        return source!!.read(buffer)
    }

    @Throws(IOException::class)
    override fun read(buffer: ByteArray, offset: Int, count: Int): Int {
        return source!!.read(buffer, offset, count)
    }

    @Throws(IOException::class)
    override fun skip(amount: Long): Long {
        return source!!.skipBytes(amount.toInt()).toLong()
    }

    override fun available(): Long {
        return try {
            source!!.length() - source!!.filePointer
        } catch (e: IOException) {
            0
        }
    }

    override fun close() {
        val src = source ?: return
        try {
            src.close()
        } catch (err: IOException) {
            // nothing to do
        }
        source = null
    }

    override fun isClosed(): Boolean = source == null

    @Throws(IOException::class)
    override fun rewind() {
        source!!.seek(0)
    }

    override fun canRewind(): Boolean = true

    override fun canRead(): Boolean = true

    override fun canWrite(): Boolean = true

    override fun canSeek(): Boolean = true

    override fun canSetLength(): Boolean = true

    @Throws(IOException::class)
    override fun write(value: Byte) {
        source!!.write(value.toInt())
    }

    @Throws(IOException::class)
    override fun write(buffer: ByteArray) {
        source!!.write(buffer)
    }

    @Throws(IOException::class)
    override fun write(buffer: ByteArray, offset: Int, count: Int) {
        source!!.write(buffer, offset, count)
    }

    @Throws(IOException::class)
    override fun setLength(length: Long) {
        source!!.setLength(length)
    }

    @Throws(IOException::class)
    override fun seek(offset: Long) {
        source!!.seek(offset)
    }

    @Throws(IOException::class)
    override fun length(): Long {
        return source!!.length()
    }
}
